tiledb/sm/query/query.cc

```diff
@@ -1024,7 +1097,45 @@ Status Query::init() {
     }
 
     RETURN_NOT_OK(check_buffer_names());
-    RETURN_NOT_OK(create_strategy());
+
+    // Cache if dimension label queries are needed.
+    bool uses_labels = uses_dimension_labels();
+    bool only_labels = uses_labels && only_dim_label_query();
+
+    // Create dimension label queries and remove labels from subarray.
+    if (uses_labels) {
+      // Initialize the dimension label queries.
+      dim_label_queries_ = tdb_unique_ptr<ArrayDimensionLabelQueries>(tdb_new(
+          ArrayDimensionLabelQueries,
+          storage_manager_,
+          array_,
+          subarray_,
+          label_buffers_,
+          buffers_,
+          fragment_name_));
+
+      // Clear the label ranges from the subarray. They are no longer needed.
+      subarray_.remove_label_ranges();
+
+      // Do not allow reading data from dimension labels for sparse arrays
+      // unless we are only returning the dimension label data. If returning
+      // data for the entire sparse array, the output for the dimension label
+      // data needs to be re-arranged to match the coordinate form we use for
+      // sparse reads. This has not yet been implemented.
+      if (!only_labels && type_ == QueryType::READ && !array_schema_->dense() &&
+          dim_label_queries_->has_data_query()) {
+        return logger_->status(Status_QueryError(
+            "Cannot initialize query; Reading dimension label data is not yet "
+            "supported on sparse arrays."));
+      }
+    }
+
+    // Create the query strategy if possible. May need to wait for range queries
+    // to be completed and for the subarray to be updated.
+    if (!uses_labels ||
+        (!only_labels && dim_label_queries_->completed_range_queries())) {
+      RETURN_NOT_OK(create_strategy());
```

Comment:
As we discussed on our call. Please skip strategy creation only for only label queries.